src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/ParameterMetaDataCacheTest.java

```diff
@@ -0,0 +1,160 @@
+            populateCharNormalCase(charValues);
+            populateNumeric(numericValues);
+            int hitsBefore = cacheHits.getInt(Class.forName("com.microsoft.sqlserver.jdbc.ParameterMetaDataCache"));
+            populateCharNormalCase(charValues);
+            int hitsAfter = cacheHits.getInt(Class.forName("com.microsoft.sqlserver.jdbc.ParameterMetaDataCache"));
+            
+            assertTrue((hitsAfter - hitsBefore) == 1);
+            con.close();
+        }
+    }
+
+    /**
+     * 
+     * Tests trimming of the cache by setting the maximum size, and size above which to trim, to 0. When the second,
+     * NUMERIC_TABLE, is updated, this should successfully trim the cache to empty, before adding the new entry.
+     * 
+     * @throws Exception
+     */
+    @Test
+    @Tag(Constants.xSQLv11)
+    @Tag(Constants.xSQLv12)
+    @Tag(Constants.xSQLv14)
+    @Tag(Constants.reqExternalSetup)
+    public void testParameterMetaDataCacheTrim() throws Exception {
+        Field cacheSize = Class.forName("com.microsoft.sqlserver.jdbc.ParameterMetaDataCache")
+                .getDeclaredField("CACHE_SIZE");
+        Field maximumWeightedCapacity = Class.forName("com.microsoft.sqlserver.jdbc.ParameterMetaDataCache")
+                .getDeclaredField("MAX_WEIGHTED_CAPACITY");
+
+        cacheSize.setAccessible(true);
+        maximumWeightedCapacity.setAccessible(true);
+
+        cacheSize.set(cacheSize.get(Class.forName("com.microsoft.sqlserver.jdbc.ParameterMetaDataCache")), 0);
+        maximumWeightedCapacity.set(
+                maximumWeightedCapacity.get(Class.forName("com.microsoft.sqlserver.jdbc.ParameterMetaDataCache")), 0);
+
+        try (SQLServerConnection con = PrepUtil.getConnection(AETestConnectionString);
+                SQLServerStatement stmt = (SQLServerStatement) con.createStatement()) {
+            String[] charValues = createCharValues(false);
+            String[] numericValues = createNumericValues(false);
+            TestUtils.dropTableIfExists(CHAR_TABLE_AE, stmt);
+            TestUtils.dropTableIfExists(NUMERIC_TABLE_AE, stmt);
+            createTable(CHAR_TABLE_AE, cekAkv, charTable);
+            createTable(NUMERIC_TABLE_AE, cekAkv, numericTable);
+
+            populateCharNormalCase(charValues);
+            populateNumeric(numericValues);
+            populateCharNormalCase(charValues);
+            con.close();
+        }
+    }
+
+    /**
+     * 
+     * Tests that the enclave is retried when using secure enclaves (assuming the server supports this). This is done by
+     * executing a query generating metadata in the cache, changing the column encryption type to make the metadata
+     * stale, and running the query again. The query should fail, but retry and pass.
+     * 
+     * @throws Exception
+     */
+    @Test
+    @Tag(Constants.xSQLv11)
+    @Tag(Constants.xSQLv12)
+    @Tag(Constants.xSQLv14)
+    @Tag(Constants.reqExternalSetup)
+    @Tag(Constants.AEv2)

```

Comment:
Similar to above, this will only work when the server supports retry (AEv3) which are enclave servers do not, so the tag is also set here to exclude.